utils/pattern/src/interpolator/mod.rs

```diff
@@ -0,0 +1,359 @@
+///     allow_raw_letters: true
+/// }).try_into().unwrap();
+///
+/// let replacements = vec![
+///     Element::Value(5),
+/// ];
+///
+/// let mut interpolator = Interpolator::new(&pattern, &replacements);
+///
+/// let mut result = vec![];
+///
+/// while let Some(element) = interpolator.try_next().expect("Failed to advance iterator") {
+///     result.push(element);
+/// }
+///
+/// assert_eq!(result, &[
+///     InterpolatedKind::Element(&Element::Value(5)),
+///     InterpolatedKind::Literal(&" days ago".into()),
+/// ]);
+/// ```
+///
+/// # Type parameters
+///
+/// - `R`: A replacement provider type implementing [`ReplacementProvider`].
+/// - `E`: An element type returned by the [`ReplacementProvider`].
+///
+/// # Lifetimes
+///
+/// - `i`: The life time of an input pattern slice.
+/// - `p`: The life time of an input [`PatternToken`], which is the life time of the string slice.
+///
+/// # Element & Replacement Provider
+///
+/// In order to allow for wide range of inputs to be interpolated using the placeholder pattern,
+/// the `Element` and [`ReplacementProvider`] types are generic.
+/// This allows the consumer of the API to decide what elements the pattern should return and how
+/// will they be identified based on any type of key that can be parsed out of a string slice.
+///
+/// This design allows for the interpolator to remain agnostic and flexible and handles wide range
+/// of ownership and life time models.
+///
+/// To simplify the common use cases, the [`ReplacementProvider`] comes with implementations for
+/// [`Vec`] (where  the placehoder key is [`usize`]) and [`HashMap`] (where the placeholder key is
+/// [`String`]) but the consumer is free to implement their own providers for any type they wish.
+///
+/// # Design Decisions
+///
+/// The interpolator is written in an intentionally generic way to enable use against wide range
+/// of potential placeholder pattern models and use cases.
+///
+/// ## Fallible Iterator
+///
+/// Rust providers a strong support for iterators and iterator combinators, which
+/// fits very well into the design of this parser/interpolator model.
+///
+/// Unfortunately, Rust iterators at the moment are infallible, while parsers are inhereantely
+/// fallible. As such, the decision has been made to design the API in line with what
+/// we hope will become a trait signature of a fallible iterator in the future, rather
+/// than implementing a reversed infallible iterator (where the [`Item`] would be
+/// `Option<Result<Item>>`).
+///
+/// Since the interpolator chains on top of the [`Parser`] it inherits the same fallible
+/// iterator API and behavior.
+///
+/// [`Item`]: std::iter::Iterator::Item
+/// [`HashMap`]: std::collections::HashMap
+/// [`Parser`]: crate::parser::Parser
+/// [`IntoIterVec`]: crate::pattern::IntoIterVec
+pub struct Interpolator<'i, 'p, R, E>
+where
+    R: ReplacementProvider<'i, E>,
+{
+    pattern: &'i Pattern<'p, R::Key>,
```

Comment:
Suggestion: Make this a `&'i [PatternToken<'s, R::Key>]` (array slice), such that you can construct an Interpolator without having to allocate a `Vec`.  We only care that it's a random-indexable list of tokens, not that it's contained in a `Vec`.